libverify/verify_libfuzzer.cpp

```diff
@@ -0,0 +1,834 @@
+/**
+ * Defines an interactive implementation of assume for libfuzzer. When
+ * assumption fails, the reason is logged, and complete the requirements
+ * @date 2019
+ */
+
+#include "verify.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <stdint.h>
+#include <stddef.h>
+#include <setjmp.h>
+#include <assert.h>
+
+/**
+ * Defines an interface to libverify, with reconfigurable data types. These
+ * types are controlled at build time using preprocessor definitions.
+ */
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*global variable Env record the environment*/
+static jmp_buf Env;
+
+/*global variable exception_type 
+ * record the return value of setjmp function,
+ * which can label different exception types*/
+static int exception_type;
+
+/*set up the exploration with Env environment,
+ * return the result of setjmp function*/
+int SetupExploration (void);
+
+/*terminate the exploration with Env environment 
+ * and different exception types, 
+ * which can jump to setjmp with different results*/
+void TerminateExploration (int exception_type);
+
+/*
+ * assign the global array with Data array*/
+void ran(const uint8_t *Data, size_t Size);
+
+/* whether the global array run out*/
+void whether_RunOut(void);
+
+/*get the next RandByte*/
+uint8_t tryGetNextRandByte();
+
+/*
+ * input the Data */
+int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size);
+
+#ifdef __cplusplus
+}
+#endif
+void sol_assert(sol_raw_uint8_t _cond, const char* _msg)
+{
+    if( !_cond){
+		fprintf(stderr,"assert");
+		if(_msg) fprintf(stderr,":%s", _msg);
+		fprintf(stderr,"\n");    
+    }
+        assert(_cond);
+}
+
+void sol_require(sol_raw_uint8_t _cond, const char* _msg)
+{
+	if( !_cond){
+		fprintf(stderr,"require");
+		if(_msg) fprintf(stderr,":%s", _msg);
+		fprintf(stderr,"\n");
+    }
+		ll_assume(_cond);	
+}
+
+void ll_assume(sol_raw_uint8_t _cond){
+    if(!_cond){
+
+	/*exception_type 2: 
+	 * the ll_assume is failure*/
+	exception_type = 2;
+	TerminateExploration(exception_type);	
+    }
+}
+
+/*
+ * declare the global array RandData 
+ * with SizeOfRandData size, 
+ * global Counter CounterOfRandData */
+uint8_t *RandData;
+size_t SizeOfRandData = 0;
+size_t CounterOfRandData = 0;
+
+
+/*set up the exploration with Env environment,
+ * return the result of setjmp function*/
+int SetupExploration (void){
+	int r;
+	r = setjmp(Env);
+	return r;
+}
+
+/*terminate the exploration with Env environment 
+ * and different exception types, 
+ * which can jump to setjmp with different results*/
+void TerminateExploration (int exception_type){
```

Comment:
Is there a reason this accepts `exception_type` as an argument? If we will never call `void TerminateExploration(int)` without passing the global, it might be simpler to change this to `void TerminateExploration()`.

Reply:
On second though, if you will always call,

```
exception_type = n;
TerminateExploration(exception_type);
```

it would be safer to call `TerminateExploration(n);`, with `TerminateException` rewritten as,

```
void TerminateExploration (int e){
    exception_type = e;
    longjmp(Env, e);
}
```